### PR TITLE
Fix for #1416 - Remove all uses of ITypeActivator and use the static method instead

### PR DESCRIPTION
**Pull request by kirthik**
Changed all CreateInstance methods to CreateFactory delegate

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/DefaultControllerFactory.cs [hunk 1]
```diff
@@ -10,15 +12,14 @@ namespace Microsoft.AspNet.Mvc
     public class DefaultControllerFactory : IControllerFactory
     {
         private readonly IServiceProvider _serviceProvider;
-        private readonly ITypeActivator _typeActivator;
         private readonly IControllerActivator _controllerActivator;
+        private static ConcurrentDictionary<Type, ObjectFactory> _controllerCache =
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/DefaultControllerFactory.cs]: this deserves a better name... `_controllerCreatorCache`?

File: src/Microsoft.AspNet.Mvc.Core/DefaultControllerFactory.cs [hunk 2]
```diff
@@ -33,9 +34,11 @@ public object CreateController(ActionContext actionContext)
                     nameof(actionContext));
             }
 
-            var controller = _typeActivator.CreateInstance(
-                _serviceProvider,
-                actionDescriptor.ControllerTypeInfo.AsType());
+            var controllerType = actionDescriptor.ControllerTypeInfo.AsType();
+
+            var fact = _controllerCache.GetOrAdd(controllerType,
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/DefaultControllerFactory.cs]: don't abbreviate names

File: src/Microsoft.AspNet.Mvc.Core/DefaultControllerFactory.cs [hunk 3]
```diff
@@ -33,9 +34,11 @@ public object CreateController(ActionContext actionContext)
...
+                ActivatorUtilities.CreateFactory(controllerType, Type.EmptyTypes));
+            var controller = fact(_serviceProvider, null);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/DefaultControllerFactory.cs]: This code isn't correct, it's not really caching anything for you.

```
var fact = _controllerCache.GetOrAdd(controllerType, ActivatorUtilities.CreateFactory(controllerType,     Type.EmptyTypes));
```

The reason is that both arguments to `GetOrAdd` must be evaluated before calling the function. That means that `CreateFactory` is going to run every call.

Here's what you can do instead.

```
var fact = _controllerCache.GetOrAdd(
    controllerType, 
    (t) => ActivatorUtilities.CreateFactory(t, Type.EmptyTypes));
```

Now you're allocating a lambda, but you'll only call `CreateFactory` once. 

But wait, there's MORE

You can do even better because the lambda is stateless.

```
private static readonly Func<Type, ObjectFactory> CreateFactory = (t) => ActivatorUtilities.CreateFactory(
    t, 
    Type.EmptyTypes)

...
...
var fact = _controllerCache.GetOrAdd(controllerType, CreateFactory);
```

Now you're guaranteed that the lambda stays static (no allocations required).

File: src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterActivator.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        public IFilter CreateInstance([NotNull] Type filterType)
+        {
+
+            var filterFact = _filterCache.GetOrAdd(filterType,
+                ActivatorUtilities.CreateFactory(filterType, Type.EmptyTypes));
+            var filter = filterFact(_provider, null) as IFilter;
+
+            if (filter == null)
+            {
+                throw new InvalidOperationException(Resources.FormatFilterFactoryAttribute_TypeMustImplementIFilter(
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterActivator.cs]: Is this string accurate? I get confused because of the name

File: src/Microsoft.AspNet.Mvc.Core/DefaultControllerFactory.cs (hunk 3 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/DefaultControllerFactory.cs]: Thanks for correcting. Will fix it.

File: src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterActivator.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        {
+            _provider = provider;
+        }
+
+        public IFilter CreateInstance([NotNull] Type filterType)
+        {
+
+            var filterFact = _filterCache.GetOrAdd(filterType,
+                ActivatorUtilities.CreateFactory(filterType, Type.EmptyTypes));
+            var filter = filterFact(_provider, null) as IFilter;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterActivator.cs]: same comments as before, don't abbreviate and fix the usage pattern.

File: src/Microsoft.AspNet.Mvc.Core/Filters/IFilterActivator.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNet.Mvc
+{
+    public interface IFilterActivator
+    {
+        IFilter CreateInstance(Type filterType);
+    }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/IFilterActivator.cs]: needs docs

File: src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultOptionActivator.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        public DefaultOptionActivator([NotNull] IServiceProvider provider)
+        {
+            _provider = provider;
+        }
+
+        public TOption CreateInstance([NotNull] Type optionType)
+        {
+            var optionFactory = _optionActivatorCache.GetOrAdd(optionType,
+                ActivatorUtilities.CreateFactory(optionType, Type.EmptyTypes));
+            return (TOption)optionFactory(_provider, null);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultOptionActivator.cs]: same comments here, fix the usage of GetOrAdd

File: src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterActivator.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    public class DefaultFilterActivator : IFilterActivator
+    {
+        private readonly IServiceProvider _provider;
+        private static readonly ConcurrentDictionary<Type, ObjectFactory> _filterCache =
+               new ConcurrentDictionary<Type, ObjectFactory>();
+
+        public DefaultFilterActivator([NotNull] IServiceProvider provider)
+        {
+            _provider = provider;
+        }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterActivator.cs]: This seems really problematic. You'd want the services here to be the request services. However, having access to request services implies that this is a cache with per-request lifetime, and that's bad for caching.

File: src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultOptionActivator.cs (hunk 7 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultOptionActivator.cs]: Sure. Will fix the abbreviated names and GetOrAdd all across and will send another PR.

File: src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterActivator.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace Microsoft.AspNet.Mvc
+{
+    public class DefaultFilterActivator : IFilterActivator
+    {
+        private readonly IServiceProvider _provider;
+        private static readonly ConcurrentDictionary<Type, ObjectFactory> _filterCache =
+               new ConcurrentDictionary<Type, ObjectFactory>();
+
+        public DefaultFilterActivator([NotNull] IServiceProvider provider)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/DefaultFilterActivator.cs]: `provider` -> either to `serviceProvider` or `services`. provider by itself is meaningless

File: src/Microsoft.AspNet.Mvc.Core/Filters/TypeFilterAttribute.cs [hunk 10]
```diff
@@ -25,18 +24,8 @@ public TypeFilterAttribute([NotNull] Type type)
...
-            if (filter == null)
-            {
-                throw new InvalidOperationException(Resources.FormatFilterFactoryAttribute_TypeMustImplementIFilter(
-                    typeof(TypeFilterAttribute).Name,
-                    typeof(IFilter).Name));
-            }
-
-            return filter;
+            var filterActivator = serviceProvider.GetRequiredService<IFilterActivator>();
+            return filterActivator.CreateInstance(ImplementationType);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/TypeFilterAttribute.cs]: Generally a `TypeFilterAttribute` is going to have application-scope. There's no need for a separate cache

File: src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultOptionActivator.cs [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    public class DefaultOptionActivator<TOption> : IOptionActivator<TOption>
+    {
+        private readonly IServiceProvider _provider;
+        private static readonly ConcurrentDictionary<Type, ObjectFactory> _optionActivatorCache =
+               new ConcurrentDictionary<Type, ObjectFactory>();
+
+        public DefaultOptionActivator([NotNull] IServiceProvider provider)
+        {
+            _provider = provider;
+        }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultOptionActivator.cs]: Same comment as FilterActivator, I'm concerned about how this is going to work with scoping

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentFactory.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        public DefaultViewComponentFactory([NotNull] IServiceProvider provider)
+        {
+            _provider = provider;
+        }
+
+        public object CreateInstance([NotNull] Type componentType)
+        {
+            var viewComponentFactory = _viewComponentCache.GetOrAdd(componentType,
+                ActivatorUtilities.CreateFactory(componentType, Type.EmptyTypes));
+            return viewComponentFactory(_provider, null);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentFactory.cs]: same comments as other factories

File: src/Microsoft.AspNet.Mvc/MvcServices.cs [hunk 13]
```diff
@@ -89,6 +88,8 @@ public static IEnumerable<IServiceDescriptor> GetDefaultServices(IConfiguration
             yield return describe.Transient<IOutputFormattersProvider, DefaultOutputFormattersProvider>();
             yield return describe.Instance<JsonOutputFormatter>(new JsonOutputFormatter());
 
+            yield return describe.Transient<IModelBinderActivator, DefaultModelBinderActivator>();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc/MvcServices.cs]: I know we always say 'transient by default' but services that have a cache of their own is enough to make this a non-default case. 

Registering a transient that:
1). DIs the service provider (which one?? it's a mystery)
2). caches data 

Seems really reckless.

File: src/Microsoft.AspNet.Mvc.Core/Filters/FilterCollectionExtensions.cs [hunk 14]
```diff
@@ -37,7 +37,7 @@ public static class FilterCollectionExtensions
         /// <param name="order">The order of the added filter.</param>
         /// <returns>An <see cref="IFilter"/> representing the added type.</returns>
         /// <remarks>
-        /// Filter instances will be created using <see cref="Microsoft.Framework.DependencyInjection.ITypeActivator"/>.
+        /// Filter instances will be created using <see cref="IFilterActivator"/>.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/FilterCollectionExtensions.cs]: No totally accurate anymore

File: src/Microsoft.AspNet.Mvc.Core/Filters/TypeFilterAttribute.cs [hunk 15]
```diff
@@ -25,18 +24,7 @@ public TypeFilterAttribute([NotNull] Type type)
...
-            var filter = obj as IFilter;
-            if (filter == null)
-            {
-                throw new InvalidOperationException(Resources.FormatFilterFactoryAttribute_TypeMustImplementIFilter(
-                    typeof(TypeFilterAttribute).Name,
-                    typeof(IFilter).Name));
-            }
-
-            return filter;
+            return (IFilter)ActivatorUtilities.CreateInstance(serviceProvider, ImplementationType);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/TypeFilterAttribute.cs]: You can do the create-factory and store it as a field on this class.

File: src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultOptionActivator.cs [hunk 16]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using Microsoft.Framework.DependencyInjection;
+
+namespace Microsoft.AspNet.Mvc.OptionDescriptors
+{
+    public class DefaultOptionActivator<TOption> : IOptionActivator<TOption>
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultOptionActivator.cs]: docs

File: src/Microsoft.AspNet.Mvc.Core/DefaultValidationExcludeFiltersProvider.cs [hunk 17]
```diff
@@ -17,13 +16,12 @@ public class DefaultValidationExcludeFiltersProvider
         /// Initializes a new instance of the DefaultBodyValidationExcludeFiltersProvider class.
         /// </summary>
         /// <param name="options">An accessor to the <see cref="MvcOptions"/> configured for this application.</param>
-        /// <param name="typeActivator">An <see cref="ITypeActivator"/> instance used to instantiate types.</param>
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/DefaultValidationExcludeFiltersProvider.cs]: Docs for all of these need to be updated, not just have `typeActivator` taken out

File: src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultModelValidatorProviderProvider.cs [hunk 18]
```diff
@@ -17,14 +16,13 @@ public class DefaultModelValidatorProviderProvider : OptionDescriptorBasedProvid
         /// Initializes a new instance of the <see cref="DefaultModelValidatorProviderProvider"/> class.
         /// </summary>
         /// <param name="options">An accessor to the <see cref="MvcOptions"/> configured for this application.</param>
-        /// <param name="typeActivator">An <see cref="ITypeActivator"/> instance used to instantiate types.</param>
-        /// <param name="serviceProvider">A <see cref="IServiceProvider"/> instance that retrieves services from the
+        /// <param name="optionActivator">A <see cref="IServiceProvider"/> instance that retrieves services from the
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultModelValidatorProviderProvider.cs]: Oopsies

File: src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultValueProviderFactoryProvider.cs [hunk 19]
```diff
@@ -17,14 +16,13 @@ public class DefaultValueProviderFactoryProvider :
         /// Initializes a new instance of the <see cref="DefaultValueProviderFactoryProvider"/> class.
         /// </summary>
         /// <param name="options">An accessor to the <see cref="MvcOptions"/> configured for this application.</param>
-        /// <param name="typeActivator">An <see cref="ITypeActivator"/> instance used to instantiate types.</param>
-        /// <param name="serviceProvider">A <see cref="IServiceProvider"/> instance that retrieves services from the
+        /// <param name="optionActivator">A <see cref="IServiceProvider"/> instance that retrieves services from the
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultValueProviderFactoryProvider.cs]: nope

File: src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/IOptionActivator.cs [hunk 20]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNet.Mvc.OptionDescriptors
+{
+    public interface IOptionActivator<TOption>
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/IOptionActivator.cs]: docs

File: src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorBasedProvider.cs [hunk 21]
```diff
@@ -15,16 +15,16 @@ namespace Microsoft.AspNet.Mvc.OptionDescriptors
     public abstract class OptionDescriptorBasedProvider<TOption>
     {
         private readonly IEnumerable<OptionDescriptor<TOption>> _optionDescriptors;
-        private readonly ITypeActivator _typeActivator;
+        private IOptionActivator<TOption> _optionActivator;
         private readonly IServiceProvider _serviceProvider;
 
         public OptionDescriptorBasedProvider(
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorBasedProvider.cs]: Add docs if you're touching something

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentFactory.cs [hunk 22]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Concurrent;
+using Microsoft.Framework.DependencyInjection;
+
+namespace Microsoft.AspNet.Mvc
+{
+    public class DefaultViewComponentFactory : IViewComponentFactory
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentFactory.cs]: docs

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentFactory.cs [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        private readonly IServiceProvider _provider;
+        private static readonly Func<Type, ObjectFactory> CreateFactory =
+            (t) => ActivatorUtilities.CreateFactory(t, Type.EmptyTypes);
+        private static readonly ConcurrentDictionary<Type, ObjectFactory> _viewComponentCache =
+               new ConcurrentDictionary<Type, ObjectFactory>();
+
+        public DefaultViewComponentFactory([NotNull] IServiceProvider provider)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentFactory.cs]: this isn't good. This is another case where caching can't work. We shouldn't see serviceprovider in the constructor of anything that performs caching.

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentFactory.cs [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+using Microsoft.Framework.DependencyInjection;
+
+namespace Microsoft.AspNet.Mvc
+{
+    public class DefaultViewComponentFactory : IViewComponentFactory
+    {
+        private readonly IServiceProvider _provider;
+        private static readonly Func<Type, ObjectFactory> CreateFactory =
+            (t) => ActivatorUtilities.CreateFactory(t, Type.EmptyTypes);
+        private static readonly ConcurrentDictionary<Type, ObjectFactory> _viewComponentCache =
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentFactory.cs]: please do a pass and get rid of static caches

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentInvokerProvider.cs [hunk 25]
```diff
@@ -9,16 +9,16 @@ namespace Microsoft.AspNet.Mvc
     public class DefaultViewComponentInvokerProvider : IViewComponentInvokerProvider
     {
         private readonly IServiceProvider _serviceProvider;
-        private readonly ITypeActivator _typeActivator;
+        private readonly IViewComponentFactory _viewComponentFactory;
         private readonly IViewComponentActivator _viewComponentActivator;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentInvokerProvider.cs]: Please rebase on top of pranav's PR and make the design the same as it is for controllers. If you think this is too big for this PR let's chat.

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/IViewComponentFactory.cs [hunk 26]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNet.Mvc
+{
+    public interface IViewComponentFactory
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/IViewComponentFactory.cs]: docs

File: src/Microsoft.AspNet.Mvc.ModelBinding/BindAttribute.cs [hunk 27]
```diff
@@ -110,11 +110,9 @@ string IModelNameProvider.Name
                 if (!initialized)
                 {
                     var services = context.OperationBindingContext.HttpContext.RequestServices;
-                    var activator = services.GetService<ITypeActivator>();
+                    var modelBinderActivator = services.GetRequiredService<IModelBinderActivator>();
 
-                    var provider = (IPropertyBindingPredicateProvider)activator.CreateInstance(
-                        services, 
-                        predicateProviderType);
+                    var provider = (IPropertyBindingPredicateProvider)modelBinderActivator.CreateInstance(services, predicateProviderType);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/BindAttribute.cs]: The design of this could be pretty similar to that of typefilterattribute. Is `IModelBinderActivator` used elsewhere?

File: src/Microsoft.AspNet.Mvc.ModelBinding/Binders/IModelBinderActivator.cs [hunk 28]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding
+{
+    public interface IModelBinderActivator
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/IModelBinderActivator.cs]: docs

File: src/Microsoft.AspNet.Mvc.ModelBinding/Binders/DefaultModelBinderActivator.cs [hunk 29]
```diff
@@ -0,0 +1,23 @@
...
+
+using System;
+using System.Collections.Concurrent;
+using Microsoft.Framework.DependencyInjection;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding
+{
+    public class DefaultModelBinderActivator : IModelBinderActivator
+    {
+        private static readonly Func<Type, ObjectFactory> CreateFactory =
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/DefaultModelBinderActivator.cs]: Could we commonize this delegate somewhere?

File: src/Microsoft.AspNet.Mvc.Core/DefaultControllerFactory.cs [hunk 30]
```diff
@@ -2,6 +2,8 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Concurrent;
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/DefaultControllerFactory.cs]: Need to rebase. This has changed

File: src/Microsoft.AspNet.Mvc.Razor/RazorPageActivator.cs [hunk 31]
```diff
@@ -14,16 +15,18 @@ public class RazorPageActivator : IRazorPageActivator
     {
         // Name of the "public TModel Model" property on RazorPage<TModel>
         private const string ModelPropertyName = "Model";
-        private readonly ITypeActivator _typeActivator;
         private readonly ConcurrentDictionary<Type, PageActivationInfo> _activationInfo;
+        private IModelMetadataProvider _metadataProvider;
+        private static readonly ConcurrentDictionary<Type, Func<IServiceProvider, object[], object>> _viewDictionaryCache =
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPageActivator.cs]: Is this needed?

File: src/Microsoft.AspNet.Mvc.Razor/RazorPageActivator.cs [hunk 32]
```diff
@@ -14,16 +15,18 @@ public class RazorPageActivator : IRazorPageActivator
     {
         // Name of the "public TModel Model" property on RazorPage<TModel>
         private const string ModelPropertyName = "Model";
-        private readonly ITypeActivator _typeActivator;
         private readonly ConcurrentDictionary<Type, PageActivationInfo> _activationInfo;
+        private IModelMetadataProvider _metadataProvider;
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPageActivator.cs]: `readonly`

File: src/Microsoft.AspNet.Mvc.Razor/VirtualPathRazorPageFactory.cs [hunk 33]
```diff
@@ -67,7 +69,9 @@ public IRazorPage CreateInstance([NotNull] string relativePath)
                     relativeFileInfo,
                     RazorCompilationService.Compile);
 
-                var page = (IRazorPage)_activator.CreateInstance(_serviceProvider, result.CompiledType);
+                var razorPageFactory = _razorPageCache.GetOrAdd(result.CompiledType, CreateFactory);
+
+                var page = (IRazorPage)razorPageFactory(_serviceProvider, null);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/VirtualPathRazorPageFactory.cs]: @yishaigalatzer, do views need type activation for views (i.e. do they need services)? Could we go down the route of using `Activator.CreateInstance` or caching the ctor for the compiled type ?

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentInvokerProvider.cs (hunk 25 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentInvokerProvider.cs]: I think it will be too big. I got a chance to talk to @yishaigalatzer and he said we should separate it from this PR. I will follow up with you tomorrow on the exact details of whats needed here and will file a separate issue. Please let me know if you have any concerns with that.

File: src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultOptionActivator.cs (hunk 16 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/DefaultOptionActivator.cs]: Sorry for missing docs in multiple places. Will be more careful going forward.

File: src/Microsoft.AspNet.Mvc.ModelBinding/BindAttribute.cs (hunk 27 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.ModelBinding/BindAttribute.cs]: Yes, its used in `BinderTypeBasedModelBinder` and `GenericModelBinder`

File: src/Microsoft.AspNet.Mvc.Razor/RazorPageActivator.cs (hunk 31 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPageActivator.cs]: No. Will remove it.

**Comment by kirthik** [comment]
Updated the code. cc @rynowak

File: src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs [hunk 34]
```diff
@@ -0,0 +1,32 @@
...
+        private readonly ConcurrentDictionary<Type, ObjectFactory> _typeActivatorCache =
+               new ConcurrentDictionary<Type, ObjectFactory>();
+
+        /// <summary>
+        /// Creates an instance of <typeparamref name="T"/>.
+        /// </summary>
+        /// <param name="serviceProvider">A <see cref="IServiceProvider"/> instance that retrieves services from the
+        /// service collection.</param>
+        /// <param name="implementationType">The <see cref="Type"/> of the <typeparamref name="T"/> to create.</param>
+        public T CreateInstance<T>([NotNull] IServiceProvider serviceProvider, [NotNull] Type implementationType)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs]: `T` -> `TVal`

File: src/Microsoft.AspNet.Mvc.Razor/VirtualPathRazorPageFactory.cs [hunk 35]
```diff
@@ -60,10 +62,12 @@ public IRazorPage CreateInstance([NotNull] string relativePath)
                 return null;
             }
 
-            var page = (IRazorPage)_activator.CreateInstance(_serviceProvider, result.CompilationResult.CompiledType);
+            var razorPageFactory = _razorPageCache.GetOrAdd(result.CompilationResult.CompiledType,
+                _createFactory);
+            var page = (IRazorPage)razorPageFactory(_serviceProvider, null);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/VirtualPathRazorPageFactory.cs]: This would be weird - you'd be activating a view with application services, not request. Either 

1. We need to flow `ViewContext` and use that to get `RequestServices`
2. We should assume Views don't need to be type activated and should use `Activator.CreateInstance`

File: src/Microsoft.AspNet.Mvc/MvcServices.cs [hunk 36]
```diff
@@ -37,9 +37,9 @@ public static IEnumerable<IServiceDescriptor> GetDefaultServices(IConfiguration
                 typeof(INestedProviderManagerAsync<>),
                 typeof(NestedProviderManagerAsync<>));
             yield return describe.Transient<MvcMarkerService, MvcMarkerService>();
-            yield return describe.Scoped(typeof(IScopedInstance<>), typeof(ScopedInstance<>));
-
-            // Core action discovery, filters and action execution.
+			yield return describe.Singleton(typeof(ITypeActivatorCache), typeof(DefaultTypeActivatorCache));
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc/MvcServices.cs]: odd formatting.

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc/MvcServices.cs]: `DefaultControllerActivator on Line 53 needs to be made transient.

File: src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs [hunk 37]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// Encapsulates information that creates a <typeparamref name="T"/> instance.
+    /// </summary>
+    public class DefaultTypeActivatorCache : ITypeActivatorCache
+    {
+        private readonly Func<Type, ObjectFactory> _createFactory =
+            (t) => ActivatorUtilities.CreateFactory(t, Type.EmptyTypes);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs]: `type => ActivatorUtilities.CreateFactory(t, Type.EmptyTypes);`

File: src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs [hunk 38]
```diff
@@ -0,0 +1,32 @@
...
+        /// <summary>
+        /// Creates an instance of <typeparamref name="T"/>.
+        /// </summary>
+        /// <param name="serviceProvider">A <see cref="IServiceProvider"/> instance that retrieves services from the
+        /// service collection.</param>
+        /// <param name="implementationType">The <see cref="Type"/> of the <typeparamref name="T"/> to create.</param>
+        public T CreateInstance<T>([NotNull] IServiceProvider serviceProvider, [NotNull] Type implementationType)
+        {
+            var createFactory = _typeActivatorCache.GetOrAdd(implementationType, _createFactory);
+            return (T)createFactory(serviceProvider, null);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs]: Named parameter for null value (in all uses of `ObjectFactory`)

File: src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentFactory.cs [hunk 39]
```diff
@@ -0,0 +1,27 @@
...
+using Microsoft.Framework.DependencyInjection;
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// Encapsulates information that creates a ViewComponent.
+    /// </summary>
+    public class DefaultViewComponentFactory : IViewComponentFactory
+    {
+        private readonly ConcurrentDictionary<Type, ObjectFactory> _viewComponentCache =
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/ViewComponents/DefaultViewComponentFactory.cs]: Unused?

File: src/Microsoft.AspNet.Mvc/MvcServices.cs (hunk 36 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc/MvcServices.cs]: Lifetime of `VirtualPathRazorPageFactory` is scoped now. Needs to be become singleton for caching to take effect.

File: test/WebSites/RequestServicesWebSite/RequestScopedActionConstraint.cs [hunk 40]
```diff
@@ -19,8 +24,10 @@ public RequestScopedActionConstraintAttribute(string requestId)
 
         public IActionConstraint CreateInstance(IServiceProvider services)
         {
-            var activator = services.GetService<ITypeActivator>();
-            return activator.CreateInstance<Constraint>(services, _requestId);
+            var constraintType = typeof(Constraint);
+            var constraintFactory = _constraintCache.GetOrAdd(constraintType,
+                CreateFactory(constraintType, _requestId));
+            return (Constraint)constraintFactory(services, new[] { _requestId });
```
Comment (pranavkm) [inline test/WebSites/RequestServicesWebSite/RequestScopedActionConstraint.cs]: Just use `ActivatorUtilities.CreateInstance`. No need to cache in a test.

File: src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs (hunk 34 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs]: How does this add more clarity? `TVal` is also an abbreviation, which we want to avoid.

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs]: Sure, `TValue` if you'd want. The guidance I'm aware of is to avoid using `T`, `U` etc and use `TVal` if there isn't something more appropriate (like `TEntry`, `TResult` etc).

File: src/Microsoft.AspNet.Mvc.Core/DefaultValidationExcludeFiltersProvider.cs [hunk 41]
```diff
@@ -17,13 +16,14 @@ public class DefaultValidationExcludeFiltersProvider
...
+        /// <param name="typeActivatorCache">As <see cref="ITypeActivatorCache"/> instance that creates an
+        ///  instance of type <see cref="IExcludeTypeValidationFilter"/>.</param>
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/DefaultValidationExcludeFiltersProvider.cs]: `As` -> `A`

When you don't have a lot to say, just be brief.

File: src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs (hunk 37 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs]: Fixed

File: src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs (hunk 34 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/DefaultTypeActivatorCache.cs]: Changed to TInstance

**Comment by rynowak** [comment]
:shipit: !!!

**Comment by kirthik** [comment]
Thanks for reviewing. Code is checked in - 5de210f5271c385d7bae47af0f866698518fccdc
